marlowe-playground-client/src/Halogen/Monaco.purs

```diff
@@ -131,37 +140,47 @@ handleAction settings Init = do
         for_ settings.codeActionProvider $ Monaco.registerCodeActionProvider monaco languageId
         for_ settings.documentFormattingEditProvider $ Monaco.registerDocumentFormattingEditProvider monaco languageId
       editor <- liftEffect $ Monaco.create monaco element languageId
-      void $ H.modify (\s -> s { editor = Just editor })
-      void $ H.subscribe $ effectEventSource (changeContentHandler monaco editor)
+      void $ H.subscribe $ effectEventSource (changeContentHandler editor)
+      void $ H.modify (\s -> s { editor = Just editor, getModelMarkers = settings.getModelMarkers, owner = settings.owner })
       H.lift $ settings.setup editor
       model <- liftEffect $ Monaco.getModel editor
       H.raise $ TextChanged (Monaco.getValue model)
       pure unit
     Nothing -> pure unit
-  where
-  changeContentHandler monaco editor (Emitter emitter) = do
-    Monaco.onDidChangeContent editor
-      ( \_ -> do
-          model <- Monaco.getModel editor
-          let
-            v = Monaco.getValue model
-
-            markersData = settings.getModelMarkers v
-          Monaco.setModelMarkers monaco model settings.owner markersData
-          markers <- Monaco.getModelMarkers monaco model
-          emitter $ Left $ HandleChange (Monaco.getValue model) markers
-      )
-    pure $ Finalizer $ pure unit
-
-handleAction _ (HandleChange contents markers) = do
-  H.raise $ TextChanged contents
-  H.raise $ MarkersChanged markers
+
+handleAction _ (HandleChange contents) =
+  void
+    $ withEditor \editor -> do
+        { getModelMarkers, owner } <- get
+        monaco <- liftEffect Monaco.getMonaco
+        model <- liftEffect $ Monaco.getModel editor
+        let
+          v = Monaco.getValue model
+
+          markersData = getModelMarkers v
+        liftEffect $ Monaco.setModelMarkers monaco model owner markersData
+        markers <- liftEffect $ Monaco.getModelMarkers monaco model
```

Comment:
Whenever I see this many `liftEffect`s in a row, I think they need pulling up. There's a shape here in the code waiting to be discovered.

Reply:
oh darn, when I wrote this I told myself I would check how many `liftEffect`s there were in the end and shift them around, but I didn't!